### PR TITLE
sync foss changes

### PR DESCRIPTION
**Pull request by mingkun2020**
*Description of changes:*
Sync foss public repo with recent changes.

*Description of how you validated changes:*
Tested in local and success

- [x] `make pr` passes
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mndeveci — APPROVED** [review]

**Comment by jfuss** [comment]
@mingkun2020 builds are failing due to improper formatting. Can you run `make black`?

**Comment by mingkun2020** [comment]
@jfuss Just wanna do the reformatting after second reviewer's approvals. If I do reformat now it may hide the actual changes which might confuse the reviewer.

**Comment by jfuss** [comment]
@mingkun2020 Please do the reformatted and I can review. Reformatting through black shouldn't hide changes, or I am miss understanding your concern. 

I am not comfortable approving or reviewing while the builds do not work.

File: integration/helpers/client_provider.py
Comment (jfuss) [inline integration/helpers/client_provider.py:124]: Why do we have a lock? First reaction is this seems odd but don't understand why we would need to do this.

File: integration/resources/expected/single/basic_api_with_mode_update.json
Comment (jfuss) [inline integration/resources/expected/single/basic_api_with_mode_update.json:3]: Why are the hashes removed here?

File: integration/helpers/client_provider.py
Comment (mingkun2020) [reply integration/helpers/client_provider.py:124]: This client provider may call by multi threads, so we add lock here to prevent concurrency issues.

Comment (jfuss) [reply integration/helpers/client_provider.py:124]: I wonder if there is a different way but not critical for this. Thanks for explaining.

File: integration/resources/expected/single/basic_api_with_mode_update.json
Comment (jfuss) [reply integration/resources/expected/single/basic_api_with_mode_update.json:3]: So the hashes are their to help guard against changes that accidentally change how SAM hashes those values. When the hashes change, this causes new resources in the stacks and can cause impact to. We had an event where we changed how Layers hashing was done and this caused throttling for customers.

Can we add these back?

**Comment by jfuss** [comment]
Approving with the condition we add back the hashing, otherwise our tests are not validating we are not spontaneously changing how the logical id's are generated. This causes unknowing deployments on customers, so the logical ids should always be deterministic.

**Review by jfuss — APPROVED** [review]

**Pull request merged**
